filters/auth/grant.go

```diff
@@ -0,0 +1,197 @@
+		Expiry:       time.Now().Add(-time.Minute),
+	}
+
+	ctx := providerContext(f.config)
+
+	// oauth2.TokenSource implements the refresh functionality,
+	// we're hijacking it here.
+	tokenSource := f.config.GetConfig().TokenSource(ctx, token)
+	return tokenSource.Token()
+}
+
+func (f *grantFilter) refreshTokenIfRequired(c cookie) (*oauth2.Token, error) {
+	now := time.Now()
+	isAccessTokenExpired := c.isAccessTokenExpired()
+	canRefresh := !f.config.DisableRefresh && c.RefreshToken != ""
+	shouldRefresh := isAccessTokenExpired || now.After(c.RefreshAfter)
+
+	if shouldRefresh {
+		if canRefresh {
+			return f.refreshToken(c)
+		} else {
+			return nil, errExpiredToken
+		}
+	} else {
+		return &oauth2.Token{
+			AccessToken:  c.AccessToken,
+			TokenType:    "Bearer",
+			RefreshToken: c.RefreshToken,
+			Expiry:       c.Expiry,
+		}, nil
+	}
+}
+
+func (f *grantFilter) setAccessTokenHeader(req *http.Request, token string) {
+	if f.config.AccessTokenHeaderName != "" {
+		req.Header.Set(f.config.AccessTokenHeaderName, authHeaderPrefix+token)
+	}
+}
+
+func (f *grantFilter) createTokenContainer(token *oauth2.Token, tokeninfo map[string]interface{}) tokenContainer {
+	subject := ""
+	if f.config.TokeninfoSubjectKey != "" {
+		subject = tokeninfo[f.config.TokeninfoSubjectKey].(string)
+	}
+
+	tokeninfo["sub"] = subject
+
+	return tokenContainer{
+		OAuth2Token: token,
+		Subject:     subject,
+		Claims:      tokeninfo,
+	}
+}
+
+func (f *grantFilter) Request(ctx filters.FilterContext) {
+	req := ctx.Request()
+
+	c, err := extractCookie(req, f.config)
+	if err == http.ErrNoCookie {
+		loginRedirect(ctx, f.config)
+		return
+	}
+
+	token, err := f.refreshTokenIfRequired(*c)
+	if err != nil && c.isAccessTokenExpired() {
+		// Refresh failed and we no longer have a valid access token.
+		loginRedirect(ctx, f.config)
+		return
+	}
+
+	tokeninfo, err := f.config.TokeninfoClient.getTokeninfo(token.AccessToken, ctx)
+	if err != nil {
+		if err == errInvalidToken {
+			log.Errorf("Error while calling tokeninfo: %v.", err)
+		}
+		loginRedirect(ctx, f.config)
+		return
+	}
+
+	f.setAccessTokenHeader(req, token.AccessToken)
+
+	// Set token in state bag for response Set-Cookie. By piggy-backing
+	// on the OIDC token container, we gain downstream compatibility with
+	// the oidcClaimsQuery filter.
+	ctx.StateBag()[oidcClaimsCacheKey] = f.createTokenContainer(token, tokeninfo)
+
+	// Set the tokeninfo also in the tokeninfoCacheKey state bag, so we
+	// can reuse e.g. the forwardToken() filter.
+	ctx.StateBag()[tokeninfoCacheKey] = tokeninfo
+}
+
+func (f *grantFilter) Response(ctx filters.FilterContext) {
+	container, ok := ctx.StateBag()[oidcClaimsCacheKey].(tokenContainer)
+	if !ok {
+		return
+	}
+
+	req := ctx.Request()
+	c, err := CreateCookie(f.config, req.Host, container.OAuth2Token)
+	if err != nil {
+		log.Errorf("Error while generating cookie: %v.", err)
+		return
+	}
+
+	rsp := ctx.Response()
+	rsp.Header.Add("Set-Cookie", c.String())
```

Comment:
In an earlier state of the PR, we only set the cookie when there was a successful login or refresh. Currently we set it in every case when there is a valid token. Is this on purpose, as in were there any problems detected with the previous approach?

Reply:
I couldn't find the version where the cookie was set conditionally. Do you have a commit-hash for me in which this was the case so that I can investigate?

Reply:
I think it's fine like this, because we check that we have something in our statebag `ctx.StateBag()[oidcClaimsCacheKey].(tokenContainer)` that we set [here](https://github.com/zalando/skipper/pull/1609/files?short_path=f3a465c#diff-357ce891fbb16fce5b8c0e90e06a755b1bc39e7b354ae1ea9b4af027479bc89cR172). 
If we have nothing in Response we return early in the first 3 lines of the Response() function.

Reply:
We should avoid setting the cookie unnecessarily. It's not a nice thing.

The commit-hash I am talking about: df776cf6d4c5a1f2af24784657ffee1c47ac1ff1

Though I don't think we should return to that hash at all, since the code has been improved and refactored a lot since then. I have a different suggestion, see further below. But first for reference: you can see in that hash, that the cookie is initially set on successful login in the callback route, then on refresh, the statebag key `oauth-grant-token` is used, but only when there was an actual refresh.

My suggestion now is to use the oidcClaimsCacheKey as is, unchanged. But in the grant filter, on the response, let's use a separate key, that would be only present if there was a refresh, and use that key for the cookie. We may have duplicate values in the statebag, but only when there was a refresh, so this is fine. WDYT?

Reply:
Let's do that

Reply:
I added a pretty naive implementation in commit f55f82e. Would that work or did I miss some context here?

Reply:
lgtm